src/libpmemcto/cto.c

```diff
@@ -0,0 +1,832 @@
+	 * but it makes it easier to support both single-file pools and
+	 * pool sets.
+	 */
+
+	/* open pool set to check consistency and to get the mapping address */
+	if ((pcp = cto_open_noinit(path, layout, cow)) == NULL) {
+		LOG(2, "cannot open pool or pool set");
+		return NULL;
+	}
+
+	/* get the last mapping address */
+	void *mapaddr = pcp->addr;
+	LOG(4, "mapping address: %p", mapaddr);
+
+	int oerrno = errno;
+	util_poolset_close(pcp->set, DO_NOT_DELETE_PARTS);
+	errno = oerrno;
+
+	/*
+	 * open the pool once again using the mapping address as a hint
+	 *
+	 * XXX: use global PMEM map hint variable to pass mapping address
+	 * to util_map()
+	 */
+	Mmap_no_random = 1;
+	Mmap_hint = mapaddr;
+
+	if ((pcp = cto_open_noinit(path, layout, cow)) == NULL) {
+		LOG(2, "cannot open pool or pool set");
+		return NULL;
+	}
+
+	struct pool_set *set = pcp->set;
+
+	if (pcp->addr != pcp) {
+		ERR("cannot mmap at the same address: %p != %p",
+				pcp, pcp->addr);
+		goto err;
+	}
+
+	/* initialize runtime parts */
+	if (cto_runtime_init(pcp, set->rdonly, set->replica[0]->is_pmem) != 0) {
+		ERR("pool initialization failed");
+		goto err;
+	}
+
+	/* Prepare pool for jemalloc */
+	if (je_vmem_pool_create(
+			(void *)((uintptr_t)pcp + CTO_DSC_SIZE_ALIGNED),
+			set->poolsize - CTO_DSC_SIZE_ALIGNED, 0, 0) == NULL) {
+		ERR("pool creation failed");
+		util_unmap(pcp->addr, pcp->size);
+		return NULL;
+	}
+
+	util_poolset_fdclose(set);
+
+	LOG(3, "pcp %p", pcp);
+	return pcp;
+
+err:
+	LOG(4, "error clean up");
+	oerrno = errno;
+	util_poolset_close(set, DO_NOT_DELETE_PARTS);
+	errno = oerrno;
+	return NULL;
+}
+
+/*
+ * pmemcto_openU -- open an existing cto memory pool
+ */
+#ifndef _WIN32
+static inline
+#endif
+PMEMctopool *
+pmemcto_openU(const char *path, const char *layout)
+{
+	LOG(3, "path \"%s\" layout \"%s\"", path, layout);
+
+	return cto_open_common(path, layout, 0);
+}
+
+#ifndef _WIN32
+/*
+ * pmemcto_open -- open an existing log memory pool
+ */
+PMEMctopool *
+pmemcto_open(const char *path, const char *layout)
+{
+	return pmemcto_openU(path, layout);
+}
+#else
+/*
+ * pmemcto_openW -- open an existing log memory pool
+ */
+PMEMctopool *
+pmemcto_openW(const wchar_t *path, const wchar_t *layout)
+{
+	char *upath = util_toUTF8(path);
+	if (upath == NULL)
+		return NULL;
+
+	char *ulayout = NULL;
+	if (layout != NULL) {
+		ulayout = util_toUTF8(layout);
+		if (ulayout == NULL) {
+			util_free_UTF8(upath);
+			return NULL;
+		}
+	}
+
+	PMEMctopool *ret = pmemcto_openU(upath, ulayout);
+	util_free_UTF8(upath);
+	util_free_UTF8(ulayout);
+	return ret;
+}
+#endif
+
+/*
+ * pmemcto_close -- close a cto memory pool
+ */
+void
+pmemcto_close(PMEMctopool *pcp)
+{
+	LOG(3, "pcp %p", pcp);
+
+	int ret = je_vmem_pool_delete(
+			(pool_t *)((uintptr_t)pcp + CTO_DSC_SIZE_ALIGNED));
+	if (ret != 0) {
+		ERR("invalid pool handle: %p", pcp);
+		errno = EINVAL;
+		return;
+	}
+
+	/* deep flush the entire pool to persistence */
```

Comment:
might as well just msync() the entire pool here. The cost of the syscall won't be big, and the kernel will do the right thing anyway.